File: person_cards.py
import json
import logging
import os

from github import Github, GithubException, InputGitTreeElement

from db import get_all_person_cards

logger = logging.getLogger(__name__)

# Fields the user fills in manually; preserved across nightly compiles.
_USER_FIELDS = ("role", "company", "contact")
_PLACEHOLDER = "—"


def _parse_user_fields(content: str) -> dict[str, str]:
    """Extract user-managed optional fields from an existing card's frontmatter."""
    fields = {k: _PLACEHOLDER for k in _USER_FIELDS}
    in_front = False
    for line in content.splitlines():
        if line.strip() == "---":
            if not in_front:
                in_front = True
                continue
            break
        if in_front:
            for key in _USER_FIELDS:
                if line.startswith(f"{key}:"):
                    value = line[len(key) + 1:].strip()
                    if value:
                        fields[key] = value
    return fields


def _parse_notes_section(content: str) -> str:
    """Preserve the freeform ## Notes body the user may have written."""
    lines = content.splitlines()
    in_notes = False
    body: list[str] = []
    for line in lines:
        if line.strip() == "## Notes":
            in_notes = True
            continue
        if in_notes:
            if line.startswith("## "):
                break
            body.append(line)
    text = "\n".join(body).strip()
    return text if text else _PLACEHOLDER


def _card_markdown(card, user_fields: dict[str, str] | None = None, notes_body: str = _PLACEHOLDER) -> str:
    if user_fields is None:
        user_fields = {k: _PLACEHOLDER for k in _USER_FIELDS}
    log_lines = "\n".join(
        f"- **{n['date']}** — {n['context']}"
        for n in sorted(json.loads(card["notes_json"]), key=lambda x: x["date"])
    )
    role = user_fields.get("role", _PLACEHOLDER)
    company = user_fields.get("company", _PLACEHOLDER)
    contact = user_fields.get("contact", _PLACEHOLDER)
    return (
        f"---\n"
        f"name: {card['name']}\n"
        f"first_seen: {card['first_seen']}\n"
        f"last_seen: {card['last_seen']}\n"
        f"role: {role}\n"
        f"company: {company}\n"
        f"contact: {contact}\n"
        f"---\n\n"
        f"# {card['name']}\n\n"
        f"## Interaction Log\n"
        f"{log_lines}\n\n"
        f"## Notes\n"
        f"{notes_body}\n"
    )


def compile_person_cards() -> None:
    cards = get_all_person_cards()
    if not cards:
        return

    vault_path = os.environ.get("GITHUB_VAULT_PATH", "Notes")
    branch = os.environ.get("GITHUB_BRANCH", "main")
    g = Github(os.environ["GITHUB_TOKEN"])
    repo = g.get_repo(os.environ["GITHUB_REPO"])

    ref = repo.get_git_ref(f"heads/{branch}")
    parent_commit = repo.get_git_commit(ref.object.sha)
    base_tree = parent_commit.tree

    blobs = []
    for card in cards:
        safe_name = card["name"].replace(" ", "-")
        path = f"{vault_path}/People/{safe_name}.md"

        # Preserve any user-edited fields from the existing card on GitHub.
        user_fields = None
        notes_body = _PLACEHOLDER
        try:
            existing = repo.get_contents(path, ref=branch)
            existing_text = existing.decoded_content.decode()
            user_fields = _parse_user_fields(existing_text)
            notes_body = _parse_notes_section(existing_text)
        except GithubException as exc:
            if exc.status != 404:
                raise

        content = _card_markdown(card, user_fields=user_fields, notes_body=notes_body)
        blob = repo.create_git_blob(content, "utf-8")
        blobs.append(InputGitTreeElement(path=path, mode="100644", type="blob", sha=blob.sha))

    new_tree = repo.create_git_tree(blobs, base_tree)
    new_commit = repo.create_git_commit(
        "people: update person cards",
        new_tree,
        [parent_commit],
    )
    ref.edit(new_commit.sha)
    logger.info("Committed %d person card(s)", len(blobs))

File: src/personCards.ts
import { getAllPersonCards } from './db';
import { GitHubClient } from './github';
import type { Env, PersonCardRow } from './types';

const USER_FIELDS = ['role', 'company', 'contact'] as const;
const PLACEHOLDER = '—';

function parseUserFields(content: string): Record<string, string> {
  const fields: Record<string, string> = Object.fromEntries(
    USER_FIELDS.map((k) => [k, PLACEHOLDER]),
  );
  let inFront = false;
  for (const line of content.split('\n')) {
    if (line.trim() === '---') {
      if (!inFront) {
        inFront = true;
        continue;
      }
      break;
    }
    if (inFront) {
      for (const key of USER_FIELDS) {
        if (line.startsWith(`${key}:`)) {
          const value = line.slice(key.length + 1).trim();
          if (value) fields[key] = value;
        }
      }
    }
  }
  return fields;
}

function parseNotesSection(content: string): string {
  const lines = content.split('\n');
  let inNotes = false;
  const body: string[] = [];
  for (const line of lines) {
    if (line.trim() === '## Notes') {
      inNotes = true;
      continue;
    }
    if (inNotes) {
      if (line.startsWith('## ')) break;
      body.push(line);
    }
  }
  const text = body.join('\n').trim();
  return text || PLACEHOLDER;
}

function cardMarkdown(
  card: PersonCardRow,
  userFields: Record<string, string>,
  notesBody: string,
): string {
  const notes: Array<{ date: string; context: string }> = JSON.parse(card.notes_json);
  const logLines = notes
    .sort((a, b) => a.date.localeCompare(b.date))
    .map((n) => `- **${n.date}** — ${n.context}`)
    .join('\n');

  return (
    `---\n` +
    `name: ${card.name}\n` +
    `first_seen: ${card.first_seen}\n` +
    `last_seen: ${card.last_seen}\n` +
    `role: ${userFields['role'] ?? PLACEHOLDER}\n` +
    `company: ${userFields['company'] ?? PLACEHOLDER}\n` +
    `contact: ${userFields['contact'] ?? PLACEHOLDER}\n` +
    `---\n\n` +
    `# ${card.name}\n\n` +
    `## Interaction Log\n` +
    `${logLines}\n\n` +
    `## Notes\n` +
    `${notesBody}\n`
  );
}

export async function compilePersonCards(env: Env): Promise<void> {
  const cards = await getAllPersonCards(env.DB);
  if (cards.length === 0) return;

  const gh = new GitHubClient(env.GH_TOKEN, env.GH_REPO);
  const branch = env.GH_BRANCH;
  const { commitSha, treeSha } = await gh.getBranchRef(branch);

  const blobs: Array<{ path: string; mode: '100644'; type: 'blob'; sha: string }> = [];

  for (const card of cards) {
    const safeName = card.name.replace(/ /g, '-');
    const path = `People/${safeName}.md`;

    let userFields = Object.fromEntries(USER_FIELDS.map((k) => [k, PLACEHOLDER]));
    let notesBody = PLACEHOLDER;

    const existing = await gh.getFileContent(path, branch);
    if (existing) {
      userFields = parseUserFields(existing);
      notesBody = parseNotesSection(existing);
    }

    const content = cardMarkdown(card, userFields, notesBody);
    const blobSha = await gh.createBlob(content);
    blobs.push({ path, mode: '100644', type: 'blob', sha: blobSha });
  }

  const newTreeSha = await gh.createTree(blobs, treeSha);
  const newCommitSha = await gh.createCommit(
    'people: update person cards',
    newTreeSha,
    commitSha,
  );
  await gh.updateRef(branch, newCommitSha);
  console.log(`Committed ${blobs.length} person card(s)`);
}
